crates/autopilot/src/infra/blockchain/circuit_breaker.rs

```diff
@@ -0,0 +1,123 @@
+use {
+    crate::{
+        domain,
+        infra::blockchain::{
+            contracts::{deployment_address, Addresses, Error},
+            Rpc,
+        },
+    },
+    ethcontract::transaction::TransactionResult,
+    primitive_types::H160,
+    std::{sync::Arc, time::Duration},
+    url::Url,
+};
+
+// Time between Ethereum blocks: 12 seconds
+// In order to fulfill Nyquist theory about sampling rate, we need to at least
+// sample twice the give frequency, in other words, half the period
+const SETTLEMENT_CHECK_SAMPLE_RATE: Duration = Duration::from_secs(6);
+
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub struct CircuitBreaker {
+    /// The authenticator contract used for allow-listing solvers to settle.
+    authenticator: contracts::GPv2AllowListAuthentication,
+    /// The safe module that is used to provide special role to EOA.
+    authenticator_role: contracts::Roles,
+    /// The EOA that is allowed to add/remove solvers.
+    authenticator_eoa: ethcontract::Account,
+    /// List of solvers to check
+    solvers: Vec<Url>,
+}
+
+impl CircuitBreaker {
+    /// Spawns a new task in order to monitor the on-chain settlement from the
+    /// solvers
+    pub async fn spawn_monitoring_task(ethrpc: Rpc, addresses: Addresses, solvers: Vec<Url>) {
+        let Rpc { web3, chain, .. } = ethrpc;
```

Comment:
I skipped the layer in `mod.rs` and add directly here the `RPC` parsing. The reason I did it like that is that in this particular case the middle layer in `mod.rs` would be purely boilerplate.